### PR TITLE
WebGLRenderer: Enable subframe upload in copyTextureToTexture, align API to 3d version

### PR DESCRIPTION
**Pull request by gkjohnson**
Related issue: -

**Description**

Adds a Box2 "bounds" argument to `copyTextureToTexture` so we can upload just a small portion of a texture.

```js
renderer.copyTextureToTexture( box, position, src, dst, level );
```

Specifically my use case is to be able copy just a single pixel of an id texture to a render target so it can be read back to the GPU to identify the id at the given pixel (see the [EXT_mesh_features Cesium glTF extension](https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_mesh_features#feature-id-by-texture-coordinates)). With this change we can copy just a single pixel from a large source texture to a 1x1 render target and then read it back whereas previously we'd have to copy to a render target of equivalent size. As far as I know there's no other way to reliably read just a single pixel from an arbitrary image.

From my testing with reading single pixel from a 2048 x 2048 image this improves the upload and read time from ~10ms to ~2ms. If async read is used presumably the ~2ms would go down quite a bit further, as well.

<details>
  <summary>Benchmark code</summary>

```js
const quad = new FullScreenQuad( new THREE.MeshBasicMaterial( { color: 0x0000ff } ) );
const srcTex = new THREE.DataTexture( new Uint8Array( 2048 * 2048 * 4 ).fill( 255 ), 4096, 4096 );
srcTex.needsUpdate = true;

{

	const target = new THREE.WebGLRenderTarget( 1, 1 );

	// initialize target
	renderer.setRenderTarget( target );
	quad.render( renderer );
	renderer.setRenderTarget( null );

	console.time( 'READ' );

	// copy data
	const bounds = new THREE.Box2();
	bounds.min.setScalar( 0 );
	bounds.max.setScalar( 1 );
	renderer.copyTextureToTexture( bounds, bounds.min, srcTex, target.texture );

	// read data
	const result = new Uint8Array( 4 );
	renderer.readRenderTargetPixels( target, 0, 0, 1, 1, result );
	console.log( result );

	console.timeEnd( 'READ' ); // ~2ms

}

{

	const target = new THREE.WebGLRenderTarget( 2048, 2048 );

	// initialize target
	renderer.setRenderTarget( target );
	quad.render( renderer );
	renderer.setRenderTarget( null );

	console.time( 'READ' );

	// copy data
	const bounds = new THREE.Box2();
	bounds.min.setScalar( 0 );
	bounds.max.setScalar( 2048 );
	renderer.copyTextureToTexture( bounds, bounds.min, srcTex, target.texture );

	// read data
	const result = new Uint8Array( 4 );
	renderer.readRenderTargetPixels( target, 0, 0, 1, 1, result );
	console.log( result );

	console.timeEnd( 'READ' ); // ~10ms

}

```
</details>

*This contribution is funded by the [Cesium GIS Ecosystem Grant](https://cesium.com/cesium-ecosystem-grants/grant-directory/)*

### CODE REVIEW

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -2371,10 +2372,26 @@ class WebGLRenderer {
...
+				srcTexture = arguments[ 1 ];
+				dstTexture = arguments[ 2 ];
+				level = arguments[ 3 ];
+
+				const image = srcTexture.isCompressedTexture ? srcTexture.mipmaps[ level ] : srcTexture.image;
+				sourceBox = new Box2();
+				sourceBox.min.set( 0, 0 );
+				sourceBox.max.set( image.width, image.height );
+
+			}
```
Comment (gkjohnson) [inline src/renderers/WebGLRenderer.js:2391]: Presumably we'd want to remove this after 10 releases?

File: src/renderers/WebGLRenderer.js [hunk 2]
```diff
@@ -2386,24 +2403,43 @@ class WebGLRenderer {
...
 
-				_gl.texSubImage2D( _gl.TEXTURE_2D, level, position.x, position.y, width, height, glFormat, glType, srcTexture.image.data );
+				_gl.texSubImage2D( _gl.TEXTURE_2D, level, position.x, position.y, width, height, glFormat, glType, image.data );
 
 			} else {
 
 				if ( srcTexture.isCompressedTexture ) {
 
-					_gl.compressedTexSubImage2D( _gl.TEXTURE_2D, level, position.x, position.y, srcTexture.mipmaps[ 0 ].width, srcTexture.mipmaps[ 0 ].height, glFormat, srcTexture.mipmaps[ 0 ].data );
+					_gl.compressedTexSubImage2D( _gl.TEXTURE_2D, level, position.x, position.y, image.width, image.height, glFormat, image.data );
```
Comment (gkjohnson) [inline src/renderers/WebGLRenderer.js:2427]: One other discrepancy I noticed between the `copyTextureToTexture` and `copyTextureToTexture3D` implementations is that the 3D variant used "level" to index into the source textures mipmap array whereas it was always 0 in the 2d version. I've aligned to what the 3d function was doing but am happy to change it either way.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:2391]: Yes, that makes sense. It's best to add the annotation `// @deprecated, r165` so we can easier search for code that can be removed in future releases.

File: src/renderers/WebGLRenderer.js [hunk 3]
```diff
@@ -2386,24 +2403,43 @@ class WebGLRenderer {
 			_gl.pixelStorei( _gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, dstTexture.premultiplyAlpha );
 			_gl.pixelStorei( _gl.UNPACK_ALIGNMENT, dstTexture.unpackAlignment );
 
+			const unpackRowLen = _gl.getParameter( _gl.UNPACK_ROW_LENGTH );
+			const unpackImageHeight = _gl.getParameter( _gl.UNPACK_IMAGE_HEIGHT );
+			const unpackSkipPixels = _gl.getParameter( _gl.UNPACK_SKIP_PIXELS );
+			const unpackSkipRows = _gl.getParameter( _gl.UNPACK_SKIP_ROWS );
+			const unpackSkipImages = _gl.getParameter( _gl.UNPACK_SKIP_IMAGES );
+
```
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:2411]: So this block is to store existing state values so they can be restored after the copy, right?

If so, can we use the `current*` or `old*` syntax like at other places?

**Comment by Mugen87** [comment]
In general, I'm okay with this change. When adding `copyTextureToTexture()` I've originally wanted the signature as simple as possible. But I agree not having the option to define the source box is unfortunate. 

Actually, I would prefer to have `sourceBox` as an optional parameter at the end of the signature but that would mean changing `copyTextureToTexture3D()` as well.

File: src/renderers/WebGLRenderer.js (hunk 3 above)
Comment (gkjohnson) [reply src/renderers/WebGLRenderer.js:2411]: I just copied the same lines from `copyTextureToTexture3D` - do you want it changed there, as well? And which is the preferred syntax, `current*` or `old*`? I think we should pick one.

Comment (gkjohnson) [reply src/renderers/WebGLRenderer.js:2411]: Looks like `current*` is used more frequently so I'll update it to that 😁

**Comment by Mugen87** [comment]
I vote for:
```js
copyTextureToTexture( position, srcTexture, dstTexture, level, srcBox );
copyTextureToTexture3D( position, srcTexture, dstTexture, level, srcBox );
```
I  agree the grouping is not ideal but always asking the users to define the source box like in the example feels inconvenient. There are for sure use cases where users want to copy the entire source texture into the destination texture and that should work as easy as possible.

**Comment by gkjohnson** [comment]
Just updated the signatures and update both examples!

**Review by Mugen87 — APPROVED** [review]

File: src/renderers/WebGLRenderer.js [hunk 4]
```diff
@@ -2371,10 +2371,25 @@ class WebGLRenderer {
 
 		};
 
-		this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level = 0 ) {
+		this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level = 0, sourceBox = null ) {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js:2374]: I think `srcRegion` is more clear and follows the rest of the parameter names.

**Comment by mrdoob** [comment]
Or something like this?

```js
( srcCoords: Vector2 | Box2, srcTexture, dstTexture, level = 0 )
```

**Comment by gkjohnson** [comment]
> Or something like this?
> ```js
> this.copyTextureToTexture = function ( srcCoords: Vector2 | Box2, srcTexture, dstTexture, level = 0 )
> ```

`position` represents the position to copy the data into in the destination while `sourceBox` represents the region to copy from the source texture.

We can change the naming to something like the following:

```js
this.copyTextureToTexture = function ( dstPosition, srcTexture, dstTexture, level = 0, srcRegion = null )
```

Thought in my opinion something like this might make the most sense. We can change to this with backwards compatibility for 10 releases:

```js
this.copyTextureToTexture = function ( srcTexture, dstTexture, srcRegion = null, dstPosition = null, level = 0 )
```

**Comment by mrdoob** [comment]
> Though in my opinion something like this might make the most sense. We can change to this with backwards compatibility for 10 releases:
> 
> ```js
> this.copyTextureToTexture = function ( srcTexture, dstTexture, srcRegion = null, dstPosition = null, level = 0 )
> ```

That looks much nicer indeed 👌

**Comment by gkjohnson** [comment]
_Okay!_ It should be ready now

**Review by Mugen87 — APPROVED** [review]

File: src/renderers/WebGLRenderer.js [hunk 5]
```diff
@@ -2371,10 +2371,52 @@ class WebGLRenderer {
 
 		};
 
-		this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level = 0 ) {
+		this.copyTextureToTexture = function ( srcTexture, dstTexture, srcRegion = null, dstPosition = null, level = 0 ) {
+
+			// support previous signature with source box first
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js:2376]: Shouldn't it be this? 
```
// support previous signature with dstPosition first
```

Comment (gkjohnson) [reply src/renderers/WebGLRenderer.js:2376]: Oh yeah - copy paste error

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:2376]: https://github.com/mrdoob/three.js/commit/0e01940a39fab5d8d1fac0e8b10f0397fd665082

**Pull request merged**